jenkins/Hono-Release-Pipeline.groovy

```diff
@@ -0,0 +1,142 @@
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+/**
+ * Jenkins pipeline script for Hono release.
+ *
+ */
+
+node {
+    def utils = evaluate readTrusted("jenkins/Hono-PipelineUtils.groovy")
+    echo "Start"
+    properties([buildDiscarder(logRotator(artifactDaysToKeepStr: '', artifactNumToKeepStr: '', daysToKeepStr: '', numToKeepStr: '3')),
+                pipelineTriggers([cron('TZ=Europe/Berlin \n # every night between 3 and 4 AM \n H 3 * * *')])])
+    properties([buildDiscarder(logRotator(artifactDaysToKeepStr: '', artifactNumToKeepStr: '', daysToKeepStr: '', numToKeepStr: '3')), parameters([
+            string(defaultValue: '',
+                    description: "The branch to build and release from.\nExamples: \n master\n1.0.x\n1.x",
+                    name: 'BRANCH',
+                    trim: false),
+            string(defaultValue: '',
+                    description: "The version identifier to use for the artifacts built and released by this job. \nExamples:\n1.0.0-M6\n1.0.0-RC1\n2.1.0",
+                    name: 'RELEASE_VERSION',
+                    trim: false),
+            string(defaultValue: '',
+                    description: "The version identifier to use during development of the next version.\nExamples:\n2.0.0-SNAPSHOT\n1.1.0-SNAPSHOT",
+                    name: 'NEXT_VERSION',
+                    trim: false)])])
+    try {
+        checkOut()
+        setReleaseVersionAndBuild()
+        commitAndTag()
+        setNextVersion()
+        commitAndPush()
+        deploy()
+        utils.publishJavaDoc('target/site/apidocs', true)
+        currentBuild.result = 'SUCCESS'
+    } catch (err) {
+        currentBuild.result = 'FAILURE'
+        echo "Error: ${err}"
+    }
+    finally {
+        echo "Build status: ${currentBuild.result}"
+        utils.notifyBuildStatus()
+    }
+}
+
+/**
+ * Checks out the specified branch from git repo
+ *
+ */
+def checkOut() {
+    stage('Checkout') {
+        echo "Check out branch: origin/${params.BRANCH}"
+        checkout([$class                           : 'GitSCM',
+                  branches                         : [[name: "origin/${params.BRANCH}"]],
+                  doGenerateSubmoduleConfigurations: false,
+                  extensions                       : [[$class: 'WipeWorkspace'],
+                                                      [$class: 'LocalBranch']],
+                  userRemoteConfigs                : [[credentialsId: 'bebef3c5-da22-425c-8554-f62d0b3a9608', url: 'ssh://github.com/eclipse/hono.git']]])
```

Comment:
Do we need to keep the credentialsId in the code?

Reply:
During credentials configuration , it seems eclipse guys didn't set an Id and left it blank. Thereby jenkins created an ID as above. It is just an Id to identify the credentials stored in jenkins. I can parametrize it. But I don't see any real value in it. For example I would have created this ID as GitCredentialsId. During parametrization , I would set the paramter name as credentialsId and set the default value as GitCredentialsId. ($credentialsId=GitCredentialsId).

But I see some value in parametrization, if we may have in future more than one credentials in stored in jenkins and we can set the credentials Id accordingly.

Reply:
I do not really understand. Is this just a logical identifier which we can make up ourselves or does this have to match the identifier that exists on our JIPP instance (and which has been created by the Eclipse folks)?

Reply:
Credentials are stored in JIPP instance for hono by the Eclipse folks. An unique ID (by which these credentials are identified from jobs and other configuration) can be provided explicitly during storing of credentials. If left blank, then jenkins will create an ID as we see here (bebef3c5-da22-425c-8554-f62d0b3a9608)

Sample screenshot of a credentials configuration page is attached below.
![image](https://user-images.githubusercontent.com/18659982/46818560-ddc5c080-cd81-11e8-829b-4896fcae6db8.png)